src/client/QXmppHttpUploadManager.cpp

```diff
@@ -0,0 +1,255 @@
+        return upload;
+    }
+
+    if (fileSize < 0) {
+        if (!data->isSequential()) {
+            fileSize = data->size();
+        } else {
+            warning(QStringLiteral("No fileSize set and cannot determine size from IO device."));
+            upload->d->reportError({QStringLiteral("File size MUST be set for sequential devices."), std::any()});
+            upload->d->reportFinished();
+            return upload;
+        }
+    }
+
+    auto future = client()->findExtension<QXmppUploadRequestManager>()->requestSlot(filename, fileSize, mimeType);
+    await(future, this, [this, upload, data](SlotResult result) {
+        // first check whether upload was cancelled in the meantime
+        if (upload->d->cancelled) {
+            upload->d->reportFinished();
+            return;
+        }
+
+        if (std::holds_alternative<QXmppStanza::Error>(result)) {
+            upload->d->reportError(std::get<QXmppStanza::Error>(std::move(result)));
+            upload->d->reportFinished();
+        } else {
+            auto slot = std::get<QXmppHttpUploadSlotIq>(std::move(result));
+            upload->d->getUrl = slot.getUrl();
+
+            QNetworkRequest request(slot.putUrl());
+            auto headers = slot.putHeaders();
+            for (auto itr = headers.cbegin(); itr != headers.cend(); ++itr) {
+                request.setRawHeader(itr.key().toUtf8(), itr.value().toUtf8());
+            }
+            auto *reply = d->netManager->put(request, data);
+            upload->d->reply = reply;
+
+            connect(reply, &QNetworkReply::finished, this, [reply, upload]() {
+                if (reply->error() == QNetworkReply::NoError) {
+                    upload->d->reportFinished();
+                }
+                reply->deleteLater();
+            });
+
+#if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)
+            connect(reply, &QNetworkReply::errorOccurred, this,
+#else
+            connect(reply, QOverload<QNetworkReply::NetworkError>::of(&QNetworkReply::error), this,
+#endif
+                    [upload, reply](QNetworkReply::NetworkError error) {
+                upload->d->reportError({reply->errorString(), error});
+                upload->d->reportFinished();
+                reply->deleteLater();
+            });
+
+            connect(reply, &QNetworkReply::uploadProgress, this, [upload](qint64 sent, qint64 total) {
+                quint64 sentBytes = sent < 0 ? 0 : quint64(sent);
+                quint64 totalBytes = total < 0 ? 0 : quint64(total);
+                upload->d->reportProgress(sentBytes, totalBytes);
+            });
+        }
+    });
+
+    return upload;
+}
+
+std::shared_ptr<QXmppHttpUpload> QXmppHttpUploadManager::uploadFile(const QFileInfo &fileInfo, const QString &filename)
+{
+    auto file = std::make_unique<QFile>(fileInfo.absoluteFilePath());
+    if (!file->open(QIODevice::ReadOnly)) {
+        std::shared_ptr<QXmppHttpUpload> upload(new QXmppHttpUpload);
```

Comment:
why not std::make_shared?

Reply:
that doesn't work, because `make_shared` is not allowed to call the private constructor, only QXmppHttpUploadManager is allowed.